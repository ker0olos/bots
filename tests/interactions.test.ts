// deno-lint-ignore-file no-explicit-any

import {
  assertEquals,
  assertRejects,
} from 'https://deno.land/std@0.168.0/testing/asserts.ts';

import {
  assertSpyCalls,
  stub,
} from 'https://deno.land/std@0.168.0/testing/mock.ts';

import * as search from '../src/search.ts';

import {
  Character,
  CharacterRole,
  Format,
  Media,
  RelationType,
  Type,
} from '../src/types.ts';

Deno.test('media', async (test) => {
  await test.step('normal search', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
        romaji: 'romaji title',
        native: 'native title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            footer: {
              text: 'native title',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('non-english title', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        native: 'native title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'native title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('external links', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
      externalLinks: [
        { site: 'FakeTube', url: 'url' },
        { site: 'Crunchyroll', url: 'url' },
      ],
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [
            {
              type: 1,
              components: [
                {
                  url: 'url',
                  label: 'Crunchyroll',
                  style: 5,
                  type: 2,
                },
              ],
            },
          ],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('youtube trailer', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
      trailer: {
        site: 'youtube',
        id: 'video_id',
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [
            {
              type: 1,
              components: [
                {
                  url: 'https://youtu.be/video_id',
                  label: 'Trailer',
                  style: 5,
                  type: 2,
                },
              ],
            },
          ],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('characters embeds', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
      characters: {
        edges: [{
          role: CharacterRole.Main,
          node: {
            id: 5,
            name: {
              full: 'main character name',
            },
            description: 'main character description',
            image: {
              medium: 'main character url',
            },
            gender: 'Male',
            age: '69',
          },
        }, {
          role: CharacterRole.Supporting,
          node: {
            id: 5,
            name: {
              full: 'supporting character name',
            },
            description: 'supporting character description',
            image: {
              medium: 'supporting character url',
            },
          },
        }, {
          role: CharacterRole.Background,
          node: {
            id: 5,
            name: {
              full: 'background character name',
            },
            description: 'background character description',
            image: {
              medium: 'background character url',
            },
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }, {
            type: 2,
            footer: {
              text: 'Male, 69',
            },
            title: 'main character name',
            color: 16777215,
            description: 'main character description',
            thumbnail: {
              url: 'main character url',
            },
          }, {
            type: 2,
            title: 'supporting character name',
            color: 16777215,
            description: 'supporting character description',
            thumbnail: {
              url: 'supporting character url',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('media relations', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
      relations: {
        edges: [{
          relationType: RelationType.Sequel,
          node: {
            id: 5,
            type: Type.Anime,
            format: Format.TV,
            popularity: 0,
            title: {
              english: 'sequel',
            },
          },
        }, {
          relationType: RelationType.Prequel,
          node: {
            id: 10,
            type: Type.Anime,
            format: Format.TV,
            popularity: 0,
            title: {
              english: 'prequel',
            },
          },
        }, {
          relationType: RelationType.SideStory,
          node: {
            id: 15,
            type: Type.Anime,
            format: Format.TV,
            popularity: 0,
            title: {
              english: 'side story',
            },
          },
        }, {
          relationType: RelationType.Adaptation,
          node: {
            id: 20,
            type: Type.Manga,
            format: Format.Manga,
            popularity: 0,
            title: {
              english: 'adaptation',
            },
          },
        }, {
          relationType: RelationType.Other,
          node: {
            id: 25,
            type: Type.Anime,
            format: Format.TV,
            popularity: 0,
            title: {
              english: 'uninteresting relation',
            },
          },
        }, {
          relationType: RelationType.Adaptation,
          node: {
            id: 30,
            type: Type.Anime,
            format: Format.TV,
            popularity: 0,
            title: {
              english: 'second adaptation',
            },
          },
        }, {
          relationType: RelationType.Adaptation,
          node: {
            id: 35,
            type: Type.Manga,
            format: Format.Manga,
            popularity: 0,
            title: {
              english: 'third adaptation',
            },
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [
            {
              type: 1,
              components: [
                {
                  custom_id: 'media:5',
                  label: 'sequel (Sequel)',
                  style: 2,
                  type: 2,
                },
                {
                  custom_id: 'media:10',
                  label: 'prequel (Prequel)',
                  style: 2,
                  type: 2,
                },
                {
                  custom_id: 'media:15',
                  label: 'side story (Side Story)',
                  style: 2,
                  type: 2,
                },
                {
                  custom_id: 'media:20',
                  label: 'adaptation (Manga)',
                  style: 2,
                  type: 2,
                },
                {
                  custom_id: 'media:30',
                  label: 'second adaptation (Anime)',
                  style: 2,
                  type: 2,
                },
              ],
            },
          ],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('music relations', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.TV,
      description: 'long description',
      popularity: 0,
      title: {
        english: 'english title',
      },
      coverImage: {
        color: '#ffffff',
        extraLarge: 'image_url',
      },
      relations: {
        edges: [{
          relationType: RelationType.Other,
          node: {
            id: 5,
            type: Type.Anime,
            format: Format.Music,
            popularity: 0,
            title: {
              english: 'op',
            },
            externalLinks: [{ site: 'youtube', url: 'youtube url' }],
          },
        }, {
          relationType: RelationType.Other,
          node: {
            id: 10,
            type: Type.Anime,
            format: Format.Music,
            popularity: 0,
            title: {
              english: 'fk',
            },
            externalLinks: [{ site: 'spotify', url: 'spotify url' }],
          },
        }, {
          relationType: RelationType.Other,
          node: {
            id: 15,
            type: Type.Anime,
            format: Format.Music,
            popularity: 0,
            title: {
              english: 'ed',
            },
            externalLinks: [{ site: 'spiketone', url: 'spiketone url' }],
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.media({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            author: {
              name: 'Anime',
            },
            title: 'english title',
            color: 16777215,
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [
            {
              type: 1,
              components: [
                {
                  url: 'youtube url',
                  label: 'op',
                  style: 5,
                  type: 2,
                },
                {
                  url: 'spotify url',
                  label: 'fk',
                  style: 5,
                  type: 2,
                },
                {
                  url: 'spiketone url',
                  label: 'ed',
                  style: 5,
                  type: 2,
                },
              ],
            },
          ],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('not found', async () => {
    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [],
              },
            },
          })),
      } as any),
    );

    try {
      await assertRejects(
        async () => await search.media({ search: 'query' }),
        Error,
        '404',
      );

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });
});

Deno.test('character', async (test) => {
  await test.step('normal search', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      age: '420',
      gender: 'male',
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            title: 'full name',
            description: 'long description',
            image: {
              url: 'image_url',
            },
            footer: {
              text: 'Male, 420',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('gender only', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      gender: 'female',
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            title: 'full name',
            description: 'long description',
            image: {
              url: 'image_url',
            },
            footer: {
              text: 'Female',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('age only', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      age: '18+',
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            title: 'full name',
            description: 'long description',
            image: {
              url: 'image_url',
            },
            footer: {
              text: '18+',
            },
          }],
          components: [],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('relations', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      media: {
        edges: [{
          characterRole: CharacterRole.Main,
          node: {
            id: 5,
            type: Type.Anime,
            format: Format.Movie,
            popularity: 0,
            title: {
              english: 'movie',
            },
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [{
            type: 2,
            title: 'full name',
            description: 'long description',
            image: {
              url: 'image_url',
            },
          }],
          components: [{
            type: 1,
            components: [{
              custom_id: 'media:5',
              label: 'movie (Movie)',
              style: 2,
              type: 2,
            }],
          }],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('not found', async () => {
    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [],
              },
            },
          })),
      } as any),
    );

    try {
      await assertRejects(
        async () => await search.character({ search: 'query' }),
        Error,
        '404',
      );

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });
});

Deno.test('character debug', async (test) => {
  await test.step('no media', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      age: '420',
      gender: 'male',
      popularity: 1_000_000,
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query', debug: true });

      assertEquals(message.json(), {
        type: 4,
        data: {
          components: [],
          content: undefined,
          embeds: [
            {
              type: 2,
              title: 'full name',
              description: undefined,
              thumbnail: {
                url: 'image_url',
              },
              fields: [
                {
                  name: 'Id',
                  value: '1',
                },
                {
                  name: 'Rating',
                  value:
                    '<:star:1061016362832642098><:star:1061016362832642098><:star:1061016362832642098><:star:1061016362832642098><:star:1061016362832642098>',
                },
                {
                  inline: true,
                  name: 'Gender',
                  value: 'male',
                },
                {
                  inline: true,
                  name: 'Age',
                  value: '420',
                },
                {
                  inline: true,
                  name: 'Media',
                  value: 'undefined',
                },
                {
                  inline: true,
                  name: 'Role',
                  value: 'undefined',
                },
                {
                  inline: true,
                  name: 'Popularity',
                  value: '1,000,000',
                },
                {
                  name: '**WARN**',
                  value:
                    'Character not available in gacha.\nAdd at least one media to the character.',
                },
              ],
            },
          ],
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('no media nor popularity', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      age: '420',
      gender: 'male',
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query', debug: true });

      assertEquals(message.json(), {
        type: 4,
        data: {
          components: [],
          content: undefined,
          embeds: [
            {
              type: 2,
              title: 'full name',
              description: undefined,
              thumbnail: {
                url: 'image_url',
              },
              fields: [
                {
                  name: 'Id',
                  value: '1',
                },
                {
                  name: 'Rating',
                  value: 'undefined',
                },
                {
                  inline: true,
                  name: 'Gender',
                  value: 'male',
                },
                {
                  inline: true,
                  name: 'Age',
                  value: '420',
                },
                {
                  inline: true,
                  name: 'Media',
                  value: 'undefined',
                },
                {
                  inline: true,
                  name: 'Role',
                  value: 'undefined',
                },
                {
                  inline: true,
                  name: 'Popularity',
                  value: 'undefined',
                },
                {
                  name: '**WARN**',
                  value:
                    'Character not available in gacha.\nAdd at least one media to the character.',
                },
              ],
            },
          ],
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('with media', async () => {
    const character: Character = {
      id: 1,
      description: 'long description',
      name: {
        full: 'full name',
      },
      image: {
        large: 'image_url',
      },
      age: '420',
      gender: 'male',
      media: {
        edges: [{
          characterRole: CharacterRole.Main,
          node: {
            id: 5,
            type: Type.Anime,
            format: Format.TV,
            popularity: 10,
            title: {
              english: 'title',
            },
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                characters: [character],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.character({ search: 'query', debug: true });

      assertEquals(message.json(), {
        type: 4,
        data: {
          components: [],
          content: undefined,
          embeds: [
            {
              type: 2,
              title: 'full name',
              description: undefined,
              thumbnail: {
                url: 'image_url',
              },
              fields: [
                {
                  name: 'Id',
                  value: '1',
                },
                {
                  name: 'Rating',
                  value:
                    '<:star:1061016362832642098><:no_star:1061016360190222466><:no_star:1061016360190222466><:no_star:1061016360190222466><:no_star:1061016360190222466>',
                },
                {
                  inline: true,
                  name: 'Gender',
                  value: 'male',
                },
                {
                  inline: true,
                  name: 'Age',
                  value: '420',
                },
                {
                  inline: true,
                  name: 'Media',
                  value: '5',
                },
                {
                  inline: true,
                  name: 'Role',
                  value: 'Main',
                },
                {
                  inline: true,
                  name: 'Popularity',
                  value: '10',
                },
              ],
            },
          ],
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });
});

Deno.test('themes', async (test) => {
  await test.step('normal search', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.Music,
      popularity: 0,
      title: {
        english: 'title',
      },
      relations: {
        edges: [{
          relationType: RelationType.Other,
          node: {
            id: 5,
            type: Type.Anime,
            format: Format.Music,
            popularity: 0,
            title: {
              english: 'music',
            },
            externalLinks: [{ site: 'youtube', url: 'youtube url' }],
          },
        }],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      const message = await search.themes({ search: 'query' });

      assertEquals(message.json(), {
        type: 4,
        data: {
          embeds: [],
          components: [
            {
              type: 1,
              components: [
                {
                  url: 'youtube url',
                  label: 'music',
                  style: 5,
                  type: 2,
                },
              ],
            },
          ],
          content: undefined,
        },
      });

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('not found', async () => {
    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [],
              },
            },
          })),
      } as any),
    );

    try {
      await assertRejects(
        async () => await search.themes({ search: 'query' }),
        Error,
        '404',
      );

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });

  await test.step('no available themes', async () => {
    const media: Media = {
      id: 1,
      type: Type.Anime,
      format: Format.Music,
      popularity: 0,
      title: {
        english: 'title',
      },
      relations: {
        edges: [],
      },
    };

    const fetchStub = stub(
      globalThis,
      'fetch',
      () => ({
        ok: true,
        json: (() =>
          Promise.resolve({
            data: {
              Page: {
                media: [media],
              },
            },
          })),
      } as any),
    );

    try {
      await assertRejects(
        async () => await search.themes({ search: 'query' }),
        Error,
        '404',
      );

      assertSpyCalls(fetchStub, 1);
    } finally {
      fetchStub.restore();
    }
  });
});
