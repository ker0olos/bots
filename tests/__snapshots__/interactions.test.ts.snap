export const snapshot = {};

snapshot[`/help 1`] = `
{
  data: {
    attachments: [
    ],
    components: [
      {
        components: [
          {
            custom_id: "help==4=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "1/5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==1=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            custom_id: "gacha=user_id",
            label: "/gacha",
            style: 2,
            type: 2,
          },
          {
            custom_id: "now=user_id",
            label: "/now",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "1.",
        },
        description: "Before anything: You will need characters.

__You get a free pull every 30 minutes (maxed at 5 pulls)__. Use \`/now\` to check how many pulls you have and how much time is left until your pulls recharge.

__Characters are exclusive__, once a player on the server finds a character, no one else can find that character in gacha again. You will need trade to get them.

__Characters vary in ranks from 1<:smol_star:1088427421096751224>to 5<:smol_star:1088427421096751224>__ A character rating is based on their popularity (as of right now, a character rating cannot be raised).

__Gacha rates depend on multiple variables__, but generally speaking the chances of a 5<:smol_star:1088427421096751224>appearing is equal or less than 1%.

\`/pull\` allows you to do gacha pulls with no animations.
​",
        footer: {
          text: "aliases: /w, /q",
        },
        title: "\`/gacha\`",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 2`] = `
{
  data: {
    attachments: [
    ],
    components: [
      {
        components: [
          {
            custom_id: "help==0=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "2/5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==2=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            custom_id: "cstars=5=user_id",
            label: "/collection stars 5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "cstars=4=user_id",
            label: "/collection stars 4",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "2.",
        },
        description: "You will need a party to use combat features, but all combat features are still in development, so parties are only an aesthetic right now.

__You can't have more than 1 party__ or save your current party to a preset and switch between presets.

__The more your party members are familiar with each other the better they preform__. This is called synergy, five party members from the same anime are better than 5 members from a mix of different media.

\`/party view\` to see your current party.

\`/party assign name: spot:\` to assign a character you have to your party;
leaving the \`spot:\` empty will assign the character to the first empty spot or override the last spot.

\`/collection\` can be used to browse your characters and select your potential party members.
​",
        footer: {
          text: "aliases: /team, /p",
        },
        title: "\`/party\`",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 3`] = `
{
  data: {
    attachments: [
    ],
    components: [
      {
        components: [
          {
            custom_id: "help==1=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "3/5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==3=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            label: "Full Roadmap",
            style: 5,
            type: 2,
            url: "https://github.com/ker0olos/fable/issues/1",
          },
          {
            label: "GitHub",
            style: 5,
            type: 2,
            url: "https://github.com/ker0olos/fable",
          },
        ],
        type: 1,
      },
      {
        components: [
          {
            label: "Discord Support Server",
            style: 5,
            type: 2,
            url: "https://discord.gg/H69RVBxeYY",
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Roadmap",
        },
        description: "__**Releasing in the near future (5 days ~ 2 months)**__

**[Customization](https://github.com/ker0olos/fable/issues/57)**
Allow players to customize their characters including images and names.

**[Synthesis](https://github.com/ker0olos/fable/issues/76)**
Merging 1-4<:smol_star:1088427421096751224>characters to get specific grade pulls (e.g. Synthesizing five 1-star characters gives a pull where you are guaranteed a 2-star character).

**[Stealing](https://github.com/ker0olos/fable/issues/49)**
You will be able to steal characters from other users.",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 4`] = `
{
  data: {
    attachments: [
    ],
    components: [
      {
        components: [
          {
            custom_id: "help==2=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "4/5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==4=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "All Commands",
        },
        description: "- \`/gacha\` \`/w\`: _start a new gacha pull_
- \`/pull\` \`/q\`: _start a quiet gacha pull with no animation_
- \`/now\` \`/vote\` \`/daily\` \`/tu\`: _check what you can do right now_
- \`/search\` \`/anime\` \`/manga\`: _search for specific media_
- \`/character\` \`/char\` \`/im\`: _search for a specific character_

- \`/party view\` \`/team view\` \`/p view\`: _view your current party_
- \`/party assign\` \`/team assign\` \`/p assign\`: _assign a character to your party_
- \`/party swap\` \`/team swap\` \`/p swap\`: _swap characters spots with each others_
- \`/party remove\` \`/team remove\` \`/p remove\`: _remove a character from your party_

- \`/collection list\` \`/coll list\` \`/mm list\`: _list user characters in bulks_
- \`/collection stars\` \`/coll stars\` \`/mm stars\`: _view user your stars_
- \`/collection media\` \`/coll media\` \`/mm media\`: _view user characters from a specific media_

- \`/found\` \`/obtained\` \`/owned\`: _list all characters found from a specific media_


- \`/image\` \`/custom\`: _change the image of a character_
- \`/nick\`: _change the nickname of a character_

- \`/trade\` \`/offer\`: _trade characters with another user_
- \`/give\` \`/gift\`: _give characters to another user_

- \`/anilist next_episode\`: _find when the next episode of an anime is airing_",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;
